### PR TITLE
Adding support for formatting CoCo OS-9 file systems

### PR DESCRIPTION
**Pull request by npwoods**
A caveat of this support is that there is no way for a FS implementation to get the actual floppy geometry.  Therefore, we are currently hard coding the track count, head count and sector size (attn @galibert)

### CODE REVIEW

**Comment by npwoods** [comment]
I'd like for @tlindner to look at this (Tim, not sure if you're aware of this, but MAME now supports formatting the file system when creating disk images)

**Comment by tlindner** [comment]
Wow, this is handy. But I have notes:
1. The bitmap allocation isn't working correctly. I would expect a contiguous set of bits from left to right. Like this:
```FF FF FF FF 80 00 00 00 00 00 00 00 00 00 00 00```
But on a default 35 track disk I get:
```FF FF 00 80 C0 E0 F0 F8 FC FE 00 00 00 00 00 00```
2. My notes for dd_fmt in LSN0 is: bit 0 is double sided, bit 1 is double density and bit 2 is tpi. I think this only supports double density disks, bit 1 should always be set.
3. It would be nice if the creation date were set to `TODAY` in dd_dat in LSN0.

**Comment by npwoods** [comment]
@tlindner I've incorporated your requested changes; care to review again?  I even made it so that you can pass the creation date to `floptool`

I'd specifically like more info about the following:
1. DId I get the date/time conversions done correctly?  The official OS-9 documentation is somewhat ambiguous about this.
2. I'm also curious about what you know about how the format persists sector byte count.  The offsets seem to be different for both creation and reading, and furthermore they are neither documented nor contiguous with the other values.  Perhaps this is an extension, be it official or unofficial?

**Comment by tlindner** [comment]
I think you have not yet pushed the changes to github yet.

**Comment by npwoods** [comment]
You were right!

**Comment by tlindner** [comment]
The allocation algorithm is still not correct. You are calculating this to be the number of sectors you need to allocate:
```u32 total_allocated_sectors = 1 + allocation_bitmap_lsns + 1 + 7;```
I calculate that to be 10 on a 35 track disk. But the first two bytes of the bitmap are:
```FF FC```
Which add up to 14 LSNs. Also the 7 above is wrong, I see on line 521 you're setting the root directory file to be 8 sectors long, not 7.
And no, the dates are not getting set correctly. Here is how Toolshed is interrupting the LSN0 on a MAME newly created 35 track disk image:
```iMac:mame $ os9 id new5.dsk

LSN0 Information:
  Total sectors   :   630
  Track size      :   18
  Bytes in bitmap :   79
  Sectors/Cluster :   1
  Root dir sector :   2
  Disk owner      :   0.1  
  Disk attributes :   --------
  Disk ID         :   $1
  Disk format     :   $2 (48 TPI, Double Density, Single Sided)
  Sectors/track   :   18
  Boot sector     :   0
  Bootfile size   :   0
  Creation date   :   00/20/1900 00:00
  Disk name       :   UNTITLED
  Bytes/Sector    :   256
  Sectors Track 0 :   18
```
It is setting the day correct, but all of the other fields are not. I should also note the disk attributes is set to $FF by Toolshed, not 0.
Here is the file descriptor for the root directory from a MAME created disk:
```iMac:mame  $ os9 fstat new5.dsk,
File Information for new5.dsk,
  Attributes         : d-ewrewr
  Owner              : 0.0  
  Last modified date : 00/20/1900 00:57
  Link count         : 2
  File size          : 64 ($40)
  Creation date      : 00/20/1900
  Segment list:
     1. LSN1 ($3)   8 sectors
```
Both dates are wrong, and the link count is 2 (should be 1).

Also to your point #2, I don't know what you mean by "how the format persists sector byte count".

**Comment by npwoods** [comment]
I fixed _some_ of these issues, notably the date handling.

At this point I'm going to defer to you, Tim - my only source was the old `os9.cpp` imgtool driver that I wrote two decades ago; the link count defect came from that:
https://github.com/mamedev/mame/blob/master/src/tools/imgtool/modules/os9.cpp#L819

> Also to your point https://github.com/mamedev/mame/pull/2, I don't know what you mean by "how the format persists sector byte count".
This is a similar issue; the old Imgtool module but the sector byte count at offset 103 when creating an image:
https://github.com/mamedev/mame/blob/master/src/tools/imgtool/modules/os9.cpp#L761
but read it from offset 104 when reading an image:
https://github.com/mamedev/mame/blob/master/src/tools/imgtool/modules/os9.cpp#L657

This is highly likely to be incorrect behavior, and it isn't clear to me if "sector byte count" (something that is always 256 in the CoCo world) is even a "true" part of the OS-9 FS format.

Feel free to submit diffs :-)

File: src/lib/formats/fs_coco_os9.cpp [hunk 1]
```diff
@@ -377,6 +421,116 @@ void coco_os9_image::impl::drop_root_ref()
...
+	volume_header.w8(3, sectors);									// DD.TKS - track size in sectors
+	volume_header.w16b(4, (allocation_bitmap_bits + 7) / 8);		// DD.MAP - allocation bitmap in bytes
+	volume_header.w16b(6, cluster_size);							// DD.BIT - cluster size
+	volume_header.w24b(8, 1 + allocation_bitmap_lsns);				// DD.DIR - root directory LSN
+	volume_header.w16b(11, owner_id);								// DD.OWN - owner ID
+	volume_header.w8(13, attributes);								// DD.ATT - Dattributes
+	volume_header.w16b(14, disk_id);								// DD.DSK - disk ID
+	volume_header.w8(16, format_flags);								// DD.FMT - format flags
+	volume_header.w16b(17, sectors);								// DD.SPT - sectors per track
+	volume_header.w24b(26, creation_os9date);						// DD.DAT - date of creation
```
Comment (tlindner) [inline src/lib/formats/fs_coco_os9.cpp:468]: In LSN0 date/time is only 5 bytes: Y:M:D:H:M

**Comment by tlindner** [comment]
I can't find support in the official OS-9 documentation for a "Sector Byte Count" at offset $67 (103) in LSN0. I'll ask some experts I know.

**Comment by npwoods** [comment]
>I can't find support in the official OS-9 documentation for a "Sector Byte Count" at offset $67 (103) in LSN0. I'll ask some
> experts I know.

Thanks - I'm sure that somebody told me to add that 20 years ago; I'm sure it was not a halucination

**Comment by tlindner** [comment]
This looks good to me.

**Pull request merged**